### ISSUE TITLE
Creating new files is not multi-goroutine safe

### ISSUE THREAD

**Issue by sfc-gh-thardie** [go-git/go-billy#40]
If you call `New()` on the `storage` objects from multiple goroutines at the same time, go runtime may error with `fatal error: concurrent map writes`. This happens because the `files` map inside the `storage` struct has no concurrency protection around multiple access.

Other concurrency issues probably exist for accessing the `children` map inside the `storage` object.

**Comment by pjbgf** [comment]
@sfc-gh-thardie Yep, this is an issue that unfortunately ends up affecting `go-git` as well. I would support a new implementation keeping the current as-is. Then offer users a way to opt into the safer approach.

From an API perspective, we could do something similar to the approach on the recent `osfs` changes, in which we introduced sub-types [with options](https://github.com/go-git/go-billy/blob/d23351b687614587d25b860a7cf9bfb866e4aeef/osfs/os_test.go#L21-L23). Is this something you would be keen to implement?

**Comment by wmcnamee-coreweave** [comment]
I'm interested in this as well

**Comment by pjbgf** [comment]
Anyone interested on working on this feel free to propose a PR against the [branch](https://github.com/go-git/go-billy/tree/v6-exp) tracking for v6.
